### ISSUE TITLE
[Q] How to log at different frequencies and 2 different training runs

### ISSUE THREAD

**Issue by meghbhalerao** [wandb/wandb#4234]
I have the following code structure

```
for ep in range(num_epoch):
    # do some training stuff
    wandb.log({"some_metric": some_metric})
    if ep % 5 == 0:
        wandb.log({"another": another})

for ep in range(num_epoch):
    # do some different training stuff say stage 2 training
    wandb.log({"some_metric_2": some_metric_2})
```
I want to log the above metrics in such a fashion. But the problem that I face is the following - the logger is maintaining a global step counter so that is causing inconsistencies in logging which can be obviously seen from the above code. I tried instantiating different loggers by `run = wandb.init()`, but that does not seem to be working.

**Comment by lesliewandb** [comment]
Hi @meghbhalerao, you can use the step parameter in your log function: `wandb.log({'loss': 0.2}, step=step)` in order to log at different steps. Here's a docs page with more information on how to do this: https://docs.wandb.ai/guides/track/log#stepwise-and-incremental-logging

**Comment by meghbhalerao** [comment]
Hi, thanks for your help. But, I dont think this solved my issue. 
The logs also mention that the step size must be a monotonically increasing natural numbers here - https://docs.wandb.ai/guides/track/log#stepwise-and-incremental-logging - so I can't log at any given step that I want. 
Please let me know if I am missing something or if I have misunderstood something.

**Comment by lesliewandb** [comment]
You're right, we currently don't allow for non-monotonically increasing values for our step counter. Can you tell me what the use case to have this feature would be?

**Comment by meghbhalerao** [comment]
Hi @lesliewandb,
Thanks for your post and sorry for the late response. So there are multiple use cases that I have come across - 
1. For example I have 2 quantities that I want to log, say `a` and `b`. I want to log `a` at every training epoch, and say I want to log `b` every `n` iterations. It will be very useful if we can instantiation 2 wandb objects within the same process and log them at whatever times that we want. I think the implementation would be very simple like for example - 

```
run1 = wandb.init(some_params)
run2 = wandb.init(some_other_params)
```
and then I can log these 2 runs at whatever place I want using `run.log(whatever_dict)`
I need this use case for my project where I want to log the loss at every training epoch and I want to log some images which I am generating at every `1000` epochs since the image takes up a lot of space and it is not needed to store tens of thousands of things. Now, I am doing some hack like I am logging images at every `epoch % 1000` iterations using wandb.log(image_dict) and at the remaining iterations I am logging images as `None` - which is a very hacky way to do it.

3. Another use case for my project is that my machine learning training happens in multiple stages - for example, I run multiple training loops within the same python script (and I think this is very common in machine where multi stage training is done.) - in this case what happens is that suppose I have n1 epochs in the first training loop and n2 epochs in the second loop and so on, wandb logs the second training loop from n1-th iteration through n1 + n2 th iteration rather than logging from  0 to n2-th iteration - which makes it very hard for comparison when I have to compare 2 runs.

4. Another use case which commonly comes up and is also coming up in my project is when I trying to log different learning rates. Some learning rates have to naturally be logged at every iteration (for example cyclic  learning rate schedule in pytorch. - https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CyclicLR.html), while other LR schedulers need to be logged at every epoch which is in general different than the number of iterations in a general mini batch setting.

Summing up, I think the all the use cases can be solved if support can be added to instantiate different runs using `wandb.init()`and then logging those runs whenever the user wants. And I think the runs can be identified using some unique identifier which can be passed in as the argument to wand.init(). I think there is already an argument like name or run id which can serve as a unique ID for a given run (https://docs.wandb.ai/ref/python/init)

Thanks and please let me know if I am missing or if anything is unclear. I would also be happy to contribute if there is any help needed in implementing this feature.

-- Megh

**Comment by lesliewandb** [comment]
Thank you for your elaborate response @meghbhalerao! I'll make a ticket for this and let you know if there are any updates on it
